src/Orleans/Async/Cancellation/GrainCancellationToken.cs

```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        public CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        internal bool IsCancellationRequested => _cancellationTokenSource.IsCancellationRequested;
+
+        internal Task Cancel()
+        {
+            _cancellationTokenSource.Cancel();
+
+            if (Interlocked.Exchange(ref _isInCancellationProcess, 1) != 0)
+            {
+                return TaskDone.Done;
+            }
+
+            if (_targetGrainReferences.IsEmpty)
+            {
+                return TaskDone.Done;
+            }
+
+            var cancellationTasks = new List<Task>();
+
+            foreach (var pair in _targetGrainReferences)
+            {
+                Interlocked.Increment(ref _cancellationRequestsInFlight);
+                var cancellationTask = pair.Value.AsReference<ICancellationSourcesExtension>().CancelTokenSource(this);
+                cancellationTasks.Add(cancellationTask);
+                cancellationTask.ContinueWith(task =>
+                {
+                    if (Interlocked.Decrement(ref _cancellationRequestsInFlight) == 0)
+                    {
+                        Interlocked.Exchange(ref _isInCancellationProcess, 0);
+                    }
+
+                    if (task.IsFaulted) return;
+
+                    // remove reference to which cancellation call has succeded, 
+                    // in order to avoid unnecessary remote call in case of retrying
+                    GrainReference grainRef;
+                    _targetGrainReferences.TryRemove(pair.Key, out grainRef);
+                });
+            }
+
+            return Task.WhenAll(cancellationTasks);
+        }
+
+        internal void AddGrainReference(GrainReference grainReference)
+        {
+            _targetGrainReferences.AddOrUpdate(grainReference.GrainId, id => grainReference, (id, ignore) => grainReference);
+        }
+
+        public void Dispose()
+        {
+            _cancellationTokenSource.Dispose();
+        }
+
+        #region Serialization
+
+        [SerializerMethod]
+        internal static void SerializeGrainCancellationToken(object obj, BinaryTokenStreamWriter stream, Type expected)
+        {
+            var ctw = (GrainCancellationToken)obj;
+            var canceled = ctw.CancellationToken.IsCancellationRequested;
+            stream.Write(canceled);
+            stream.Write(ctw.Id);
+        }
+
+        [DeserializerMethod]
+        internal static object DeserializeGrainCancellationToken(Type expected, BinaryTokenStreamReader stream)
+        {
+            var cancellationRequested = stream.ReadToken() == SerializationTokenType.True;
+            var tokenId = stream.ReadGuid();
+            return new GrainCancellationToken(tokenId, cancellationRequested);
+        }
+
+        [CopierMethod]
+        internal static object CopyGrainCancellationToken(object obj)
+        {
+            return obj; // CancellationToken is value type
```

Comment:
But if we to deep copy the references - then the next thing will happen: for each deep copy of token that already contains the references the following cancelling will execute one more cancel call to the reference. So if token with one reference was passed 5 times inside one silo - then at least 15 calls will be issued.